### PR TITLE
CMake: Build fuzzing Expat once, not twelve times

### PR DESCRIPTION
**Pull request by hartwork**
CC @hctim

### CODE REVIEW

File: expat/CMakeLists.txt [hunk 1]
```diff
@@ -391,10 +391,14 @@ if(EXPAT_BUILD_FUZZERS)
...
     set(fuzz_targets xml_parse_fuzzer xml_parsebuffer_fuzzer)
 
+    add_library(fuzzpat STATIC ${expat_SRCS})
+    target_compile_options(fuzzpat PRIVATE -fsanitize=fuzzer-no-link)
+
     foreach(fuzz_target ${fuzz_targets})
         foreach(encoding_type ${encoding_types})
             set(target_name ${fuzz_target}_${encoding_type})
-            add_executable(${target_name} fuzz/${fuzz_target}.c ${expat_SRCS})
+            add_executable(${target_name} fuzz/${fuzz_target}.c)
```
Comment (hctim) [inline expat/CMakeLists.txt]: Isn't the previous variant only rebuilding each expat source file once? I don't see where the building six libraries comes from?

pseudocode of previous:
```
for src in expat_SRCS
  clang -fsanitize=fuzzer-no-link $src -o $src.o
  echo $src.o >> sources.txt
done
clang -fsanitize=fuzzer ${fuzz_target} $(cat sources.txt) -o ${target_name}
```

Comment (hartwork) [reply expat/CMakeLists.txt]: It's the `add_executable(… ${expat_SRCS})` that makes CMake compile every file in `${expat_SRCS}` (3 files) per encoding (6 encodings) per fuzz target (2 targets). So it's actually 12+1 times we compile Expat, not 6+1, my bad :smile: Will fix the title now.

For proof (on Git `master`):
```console
# cmake -DEXPAT_BUILD_FUZZERS=ON -DCMAKE_CXX_FLAGS=-fsanitize=address -DCMAKE_C_FLAGS=-fsanitize=address -DCMAKE_EXE_LINKER_FLAGS=-fsanitize=address -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ [..]
# make clean ; time grep -Eo 'xml_parse(buffer)?_fuzzer_[A-Z0-9-]+' Makefile | sort -u | xargs make
Scanning dependencies of target xml_parsebuffer_fuzzer_ISO-8859-1
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_ISO-8859-1.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_ISO-8859-1.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_ISO-8859-1.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_ISO-8859-1.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_ISO-8859-1
[100%] Built target xml_parsebuffer_fuzzer_ISO-8859-1
Scanning dependencies of target xml_parsebuffer_fuzzer_US-ASCII
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_US-ASCII.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_US-ASCII.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_US-ASCII.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_US-ASCII.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_US-ASCII
[100%] Built target xml_parsebuffer_fuzzer_US-ASCII
Scanning dependencies of target xml_parsebuffer_fuzzer_UTF-16
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_UTF-16
[100%] Built target xml_parsebuffer_fuzzer_UTF-16
Scanning dependencies of target xml_parsebuffer_fuzzer_UTF-16BE
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16BE.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16BE.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16BE.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16BE.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_UTF-16BE
[100%] Built target xml_parsebuffer_fuzzer_UTF-16BE
Scanning dependencies of target xml_parsebuffer_fuzzer_UTF-16LE
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16LE.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16LE.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16LE.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-16LE.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_UTF-16LE
[100%] Built target xml_parsebuffer_fuzzer_UTF-16LE
Scanning dependencies of target xml_parsebuffer_fuzzer_UTF-8
[ 20%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-8.dir/fuzz/xml_parsebuffer_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-8.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-8.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parsebuffer_fuzzer_UTF-8.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parsebuffer_fuzzer_UTF-8
[100%] Built target xml_parsebuffer_fuzzer_UTF-8
Scanning dependencies of target xml_parse_fuzzer_ISO-8859-1
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_ISO-8859-1.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_ISO-8859-1.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_ISO-8859-1.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_ISO-8859-1.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_ISO-8859-1
[100%] Built target xml_parse_fuzzer_ISO-8859-1
Scanning dependencies of target xml_parse_fuzzer_US-ASCII
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_US-ASCII.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_US-ASCII.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_US-ASCII.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_US-ASCII.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_US-ASCII
[100%] Built target xml_parse_fuzzer_US-ASCII
Scanning dependencies of target xml_parse_fuzzer_UTF-16
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_UTF-16
[100%] Built target xml_parse_fuzzer_UTF-16
Scanning dependencies of target xml_parse_fuzzer_UTF-16BE
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16BE.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16BE.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16BE.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16BE.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_UTF-16BE
[100%] Built target xml_parse_fuzzer_UTF-16BE
Scanning dependencies of target xml_parse_fuzzer_UTF-16LE
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16LE.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16LE.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16LE.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-16LE.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_UTF-16LE
[100%] Built target xml_parse_fuzzer_UTF-16LE
Scanning dependencies of target xml_parse_fuzzer_UTF-8
[ 20%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-8.dir/fuzz/xml_parse_fuzzer.c.o
[ 40%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-8.dir/lib/xmlparse.c.o
[ 60%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-8.dir/lib/xmlrole.c.o
[ 80%] Building C object CMakeFiles/xml_parse_fuzzer_UTF-8.dir/lib/xmltok.c.o
[100%] Linking C executable fuzz/xml_parse_fuzzer_UTF-8
[100%] Built target xml_parse_fuzzer_UTF-8

real    0m25.016s
user    0m22.700s
sys     0m2.307s
```
(Yes, `-j2` or more speeds this up.)

Comment (hctim) [reply expat/CMakeLists.txt]: Got it, LGTM then.

Comment (hartwork) [reply expat/CMakeLists.txt]: Cool, thank you!

**Pull request merged**